src/Adapters/Etw.Net451/EtwTelemetryModule.cs

```diff
@@ -0,0 +1,247 @@
+        }
+
+        /// <summary>
+        /// Disposes the module.
+        /// </summary>
+        /// <param name="isDisposing">Indicate if it is called by Dispose().</param>
+        protected virtual void Dispose(bool isDisposing)
+        {
+            if (this.isDisposed)
+            {
+                return;
+            }
+
+            // Mark this object as disposed even when disposing run into exception, which is not expected.
+            this.isDisposed = true;
+            if (isDisposing)
+            {
+                if (this.traceEventSession != null)
+                {
+                    this.traceEventSession.Dispose();
+                }
+            }
+        }
+
+        private void EnableProviders()
+        {
+            foreach (EtwListeningRequest request in this.Sources)
+            {
+                this.EnableProvider(request);
+            }
+        }
+
+        private void EnableProvider(EtwListeningRequest request)
+        {
+            try
+            {
+                request.Validate();
+            }
+            catch (Exception ex)
+            {
+                EventSourceListenerEventSource.Log.FailedToEnableProviders(
+                    nameof(EtwTelemetryModule),
+                    string.IsNullOrEmpty(request.ProviderName) ? request.ProviderGuid.ToString() : request.ProviderName,
+                    ex.Message);
+            }
+
+            try
+            {
+                if (request.ProviderGuid != Guid.Empty)
+                {
+                    this.EnableProvider(request.ProviderGuid, request.Level, request.Keywords);
+                }
+                else
+                {
+                    this.EnableProvider(request.ProviderName, request.Level, request.Keywords);
+                }
+            }
+            catch (Exception ex)
+            {
+                EventSourceListenerEventSource.Log.FailedToEnableProviders(
+                    nameof(EtwTelemetryModule),
+                    string.IsNullOrEmpty(request.ProviderName) ? request.ProviderGuid.ToString() : request.ProviderName,
+                    ex.Message);
+            }
+        }
+
+        private void EnableProvider(Guid providerGuid, TraceEventLevel level, ulong keywords)
+        {
+            this.traceEventSession.EnableProvider(providerGuid, level, keywords);
+            this.enabledProviderIds.Add(providerGuid);
+        }
+
+        private void EnableProvider(string providerName, TraceEventLevel level, ulong keywords)
+        {
+            this.traceEventSession.EnableProvider(providerName, level, keywords);
+            this.enabledProviderNames.Add(providerName);
+        }
+
+        private void DisableProviders()
+        {
+            foreach (Guid id in this.enabledProviderIds)
+            {
+                this.traceEventSession.DisableProvider(id);
+            }
+            foreach (string providerName in this.enabledProviderNames)
```

Comment:
stylecop will complain for closing bracket without following newline. Please run build in release as there might be more issues